Assignment 2 submission

Vaibhav-Chhipa opened the pull request.

Vaibhav-Chhipa committed:
Assignment 2 submission

Bhavik-Ubhadiya commented on UserManagement.c:
```diff
@@ -0,0 +1,186 @@
+
+
+typedef struct {
+    int id;
```
We can use unsigned int if we want to restrict negative values.

Bhavik-Ubhadiya commented on UserManagement.c:
```diff
@@ -0,0 +1,186 @@
+typedef struct {
+    int id;
+    char name[50];
+    int age;
```
We can use unsigned short as age can't be negative and unsigned short is enough to store age.

Bhavik-Ubhadiya commented on UserManagement.c:
We also need to create `users.txt` and make it part of the PR as our code is tightly coupled to that file.

Bhavik-Ubhadiya commented on UserManagement.c:
```diff
@@ -0,0 +1,186 @@
+        }
+        maxID++;
+    }
+    newUser.id = maxID;
```
If we are dynamically assigning IDs to user as per the current entries, we can create one `init` function which will create one global variable `currentId`. We can then throughout use that variable to create the new user, so we don't have to run this loop every time when creating the user.

Bhavik-Ubhadiya commented on UserManagement.c:
```diff
@@ -0,0 +1,186 @@
+    printf("Your Unique ID is: %d\n", newUser.id);
+    
+
+    if(file != NULL){
```
This whole block seems unnecessary as we are assigning unique id to the new user in above block. So, it will never be the case when `u.id` will be equal to `newUser.id`.

Bhavik-Ubhadiya commented on UserManagement.c:
```diff
@@ -0,0 +1,186 @@
+        printf("4. Delete User\n");
+        printf("5. Exit\n");
+        printf("Enter your choice: ");
+        scanf("%d", &choice);
```
Using `scanf` here can cause potential issues. For example, if you try to run the program with input `3d`, it will lead to unexpected behavior. You can explore the reason behind this. Currently you can create a helper function using `fgets` to take valid input and then call the function every time we need to take integer input.

Bhavik-Ubhadiya commented on UserManagement.c:
```diff
@@ -0,0 +1,186 @@
+        fclose(file);
+    }
+    printf("Enter Name: ");
+    scanf(" %[^\n]", newUser.name);
```
As mentioned in later comment, we should avoid using `scanf` to take inputs.

Bhavik-Ubhadiya commented on UserManagement.c:
```diff
@@ -0,0 +1,186 @@
+    printf("Enter Age: ");
+    scanf("%d", &newUser.age);
+
+    file = fopen(FILENAME, "a");
```
We can open the file in append mode in starting and let the file pointer open if we need to use it here so that we don't have to open it again.

Bhavik-Ubhadiya commented on UserManagement.c:
```diff
@@ -0,0 +1,186 @@
+
+    int id;
+    printf("Enter the user ID to update: ");
+    scanf("%d", &id);
```
As mentioned in later comment, we should avoid using scanf to take inputs.

Bhavik-Ubhadiya commented on UserManagement.c:
```diff
@@ -0,0 +1,186 @@
+    printf("Enter the user ID to update: ");
+    scanf("%d", &id);
+
+    FILE *temp = fopen("temp.txt", "w");
```
We don't need temp.txt to update user. We can explore better approach for this.

Vaibhav-Chhipa replied:
For the update and delete operations, can we handle everything in memory (load users into an array, modify or delete as needed, and then save back to the same file)?
Just wanted to confirm if this is the preferred approach before I proceed with the changes.

Bhavik-Ubhadiya replied:
Yes, we can handle this in memory only. In actual, data is stored in databases and not in file, so these both approaches are not ideal, but from assignment point of view, handling this using memory seems better approach.

Bhavik-Ubhadiya commented on UserManagement.c:
```diff
@@ -0,0 +1,186 @@
+    printf("Enter ID of user want to delete: ");
+    scanf("%d", &id);
+
+    FILE *temp = fopen("temp.txt", "w");
```
As mentioned in earlier comment, we don't need temp.txt to delete user. We can explore better approach for this.

Vaibhav-Chhipa committed:
Update UserManagement and applied feedback

Vaibhav-Chhipa commented:
Thank you for the review and suggestions! I have updated the code to address all points:

1. Unsigned types: id is now unsigned int and age is unsigned short to prevent negative values.
2. Dynamic ID assignment: Added currentId as a global variable initialized in initSystem(). New users now get unique IDs without scanning the file each time.
3. Removed unnecessary ID check: The previous block comparing u.id with newUser.id was removed since it’s no longer needed.
4. Input handling: Replaced all scanf calls with helper functions (getIntegerInput() and getStringInput()) using fgets to safely handle input and avoid unexpected behavior with invalid inputs like “3d”.
5. File handling: The file is now opened once in append mode (userFile) at program start and used throughout, avoiding repeated opens.
6. Update/Delete operations: No temporary files are used; all modifications are handled by loading users into an array and rewriting the file, which is safer and simpler.

Bhavik-Ubhadiya commented on UserManagement.c:
```diff
@@ -0,0 +1,237 @@
+#define MAX_USERS 100
+
+// Global file pointer to keep file open in append mode
+FILE *userFile = NULL; 
```
I guess you misunderstood my suggestion on [This comment](https://github.com/Vaibhav-Chhipa/Kalpvriksha_Assignment_Vaibhav_Chhipa/pull/2#discussion_r2405318788). I said that because we were opening file two times in same function. Creating global variable is definitely the wrong choice.

Vaibhav-Chhipa replied:
Thanks for the clarification! I misunderstood your point earlier. I’ve now updated the code to avoid reopening the file and removed the global variable as suggested.

Vaibhav-Chhipa committed:
Removed Global File Pointer
